oak_idl_gen_services/src/lib.rs

```diff
@@ -0,0 +1,383 @@
+//
+// Copyright 2022 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+//! This crate allows compiling flatbuffer services to Rust in `build.rs` scripts.
+//!
+//! Also see the `oak_idl_gen_structs` crate.
+//!
+//! These crates are intentionally distinct so that `oak_idl_gen_services` can use
+//! `oak_idl_gen_structs` in order to compile the flatbuffers reflection schema at compile time.
+
+#![feature(path_file_prefix)]
+#![feature(iter_intersperse)]
+
+use anyhow::Context;
+use convert_case::Casing;
+use reflection_generated::reflection::{RPCCall, Service};
+use std::process::exit;
+
+mod reflection_generated {
+    #![allow(clippy::derivable_impls, clippy::needless_borrow)]
+    #![allow(dead_code, unused_imports)]
+
+    include!(concat!(env!("OUT_DIR"), "/reflection_generated.rs"));
+}
+
+/// Compile services from the provided flatbuffer file using the `flatc` binary installed on the
+/// system.
+///
+/// Services are generated targeting the invocation-based transport from the `oak_idl` crate (i.e.
+/// not gRPC).
+///
+/// Each service method in the `rpc_service` definition must have a unique `method_id` numeric
+/// attribute associated, which is used as part of the invocation serialization process. The
+/// attribute must be declared upfront in the same file.
+///
+/// ```text
+/// namespace Test;
+///
+/// attribute "method_id";
+///
+/// table Request {}
+/// table Response {}
+///
+/// rpc_service Name {
+///     Method(Request) : Response (method_id: 42);
+/// }
```

Comment:
I'm a fan of the method_id personally to identify methods across renames & re-orders. That said, given that flatbuffer (like grpc) does not seem to employ that concept in their RPC interface definition format, I'm not 100% convinced that adding it on is a good idea: https://google.github.io/flatbuffers/flatbuffers_guide_writing_schema.html.

Might lead to surprises for people famiiar with flatbuffer/grpc, not expecting to find our own flavor of it. :)

Reply:
fair :)